exo/train/dataset.py

```diff
@@ -0,0 +1,80 @@
+#from https://github.com/ml-explore/mlx-examples
+from pathlib import Path
+import numpy as np
+import json
+from functools import partial, reduce
+def compose(*funcs):    
+  return reduce(lambda f, g: lambda x: f(g(x)), funcs, lambda x : x)
+
+def batch_with_lengths(tokens, maxlen = None):
+  lengths = [len(x) for x in tokens]
+  batch_size = len(lengths)
+  if maxlen is None:
+    maxlen = max(lengths)
+  else:
+    lengths = [min(maxlen, l) for l in lengths]
+
+  # Pad to the max length
+  batch_arr = np.zeros((batch_size, maxlen), np.int32)
+
+  for j in range(batch_size):
+    batch_arr[j, : lengths[j]] = tokens[j]
+  batch = np.array(batch_arr)
+  return batch[:, :-1], batch[:, 1:], np.array(lengths)
+
+def batch_chunk(batch_size):
+  return lambda d, i: d[i:i + batch_size]
+  
+
+def iterate_batches(dset, batch_size, train=False, uniform_length=True):
+# Shuffle indices
+  make_batch = lambda b: batch_with_lengths(b, maxlen=dset._maxlen if uniform_length else None)
+  chunk = batch_chunk(batch_size)
+  while True:
+    indices = np.arange(len(dset))
+    if train:
+      indices = np.random.permutation(indices)
+    batch = compose(make_batch, lambda i: [dset[k] for k in i], partial(chunk, indices))
+
+    # Collect batches from dataset
+    for i in range(0, len(indices) - batch_size + 1, batch_size):
+      yield batch(i)
+
+    if not train:
+      break
+
+class Dataset:
+  def __init__(self, path: Path, preprocess=lambda item: item, loadline=json.loads, metrics={}):
```

Comment:
Yup. I was initially going to extend this to be more generally applicable to other datasets and make the length masks a parameter and the dataset loader work for "any" dataset, but it became clear that this would take more design to matter and wasn't a priority at this moment